numba/cuda/tests/cudapy/test_lineinfo.py

```diff
@@ -1,42 +1,205 @@
+        self._check(foo, sig=(int32[:],), expect=False)
 
     def test_lineinfo_in_asm(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
         @cuda.jit(lineinfo=True)
         def foo(x):
             x[0] = 1
 
-        self._check(foo, sig=(types.int32[:],), expect=True)
+        self._check(foo, sig=(int32[:],), expect=True)
+
+    def test_lineinfo_maintains_error_model(self):
+        sig = (float32[::1], float32[::1])
+
+        @cuda.jit(sig, lineinfo=True)
+        def divide_kernel(x, y):
+            x[0] /= y[0]
+
+        llvm = divide_kernel.inspect_llvm(sig)
+
+        # When the error model is Python, the device function returns 1 to
+        # signal an exception (e.g. divide by zero) has occurred. When the
+        # error model is the default NumPy one (as it should be when only
+        # lineinfo is enabled) the device function always returns 0.
+        self.assertNotIn('ret i32 1', llvm)
+
+    def test_no_lineinfo_in_device_function(self):
+        # Ensure that no lineinfo is generated in device functions by default.
+        @cuda.jit
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=False)
+
+    def test_lineinfo_in_device_function(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
+        # First we define a device function / kernel pair and run the usual
+        # checks on the generated LLVM and PTX.
+
+        @cuda.jit(lineinfo=True)
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit(lineinfo=True)
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=True)
+
+        # Now we can check the PTX of the device function specifically.
+
+        ptx = caller.inspect_asm(sig)
+        ptxlines = ptx.splitlines()
+
+        # Check that there is no device function in the PTX
+
+        # A line beginning with ".weak .func" that identifies a device function
+        devfn_start = re.compile(r'^\.weak\s*\.func')
+
+        for line in ptxlines:
+            if devfn_start.match(line) is not None:
+                self.fail(f"Found device function in PTX:\n\n{ptx}")
+
+        # Scan for .loc directives taht refer to an inlined device function
+
+        loc_directive = self._loc_directive_regex()
+        found = False
+
+        for line in ptxlines:
+            if loc_directive.search(line) is not None:
+                if 'inlined_at' in line:
+                    found = True
+                    break
+
+        if not found:
+            # Join one line either side so the function as a whole is shown,
+            # i.e. including the declaration and parameter list, and the
+            # closing brace.
```

Comment:
Is this comment applicable? It looks like it's just showing the full ptx?

Reply:
Ah, not anymore - it was from before the function was (correctly) inlined.

Reply:
Now removed.